### PR TITLE
[167] Implement update endpoint for contacts

### PR DESCRIPTION
**Pull request by darokel**
### Context

- https://trello.com/c/NLKqADvh/167-add-update-endpoint-for-ucas-contacts-in-api-v2

### Changes proposed in this pull request

- Add a new endpoint to update individual contacts for a provider

### Guidance to review

- Load the rails server locally and fire a request to update a contact: `PUT 'localhost:3001/api/v2/contacts/:id`
- Assert the contact is updated.

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by darokel** [comment]
[Add update endpoint for ucas contacts in API v2.](https://trello.com/c/NLKqADvh/167-add-update-endpoint-for-ucas-contacts-in-api-v2)

**Review by defong — CHANGES_REQUESTED** [review]
failing tests

**Comment by darokel** [comment]
> failing tests

Pushed up some fixes

File: spec/serializers/api/v2/serializable_contact_spec.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  let(:contact) { build(:contact) }
+  let(:resource) { described_class.new(object: contact) }
+
+  it "sets type to contacts" do
+    expect(resource.jsonapi_type).to eq(:contacts)
+  end
+
+  subject { JSON.parse(resource.as_jsonapi.to_json) }
+
+  it { should have_type "contacts" }
```
Comment (gpeng) [inline spec/serializers/api/v2/serializable_contact_spec.rb:13]: We should (no pun intended) use `is_expected.to` rather than `should` here.

**Review by gpeng — CHANGES_REQUESTED** [review]
This looks good for this ticket ie. 'update' but I think we at least need `show` which we have a use case for in the functionality that we're implementing. We shouldn't be getting the whole provider (which is one of our slowest endpoints) to edit a contact. We also don't have create/delete which we need a product discussion about. Where do the contacts get created initially?

File: app/serializers/api/v2/serializable_provider.rb [hunk 2]
```diff
@@ -85,6 +85,7 @@ class SerializableProvider < JSONAPI::Serializable::Resource
 
       has_many :sites
       has_many :users
+      has_many :contacts
```
Comment (defong) [inline app/serializers/api/v2/serializable_provider.rb:88]: This looks wrong

https://github.com/DFE-Digital/teacher-training-api/blob/471e479c7cf7e84de5c2dd37263cf70a19b06ba1/app/serializers/api/v2/serializable_provider.rb#L50-L68


https://github.com/DFE-Digital/teacher-training-api/blob/a97b30d65a5e2f357e696d0a5f406e6d1b144fb2/app/models/contact.rb#L14-L21

It looks like its `has_one` x 5 times
https://apidock.com/rails/ActiveRecord/Associations/ClassMethods/has_one

**Review by defong — CHANGES_REQUESTED** [review]

File: app/serializers/api/v2/serializable_provider.rb (hunk 2 above)
Comment (darokel) [reply app/serializers/api/v2/serializable_provider.rb:88]: We're steering away from this approach so it's more agnostic and not have to define an attribute for every contact.

Comment (gpeng) [reply app/serializers/api/v2/serializable_provider.rb:88]: I'm going to add another ticket to remove those so it's not part of this PR.

File: spec/serializers/api/v2/serializable_contact_spec.rb (hunk 1 above)
Comment (darokel) [reply spec/serializers/api/v2/serializable_contact_spec.rb]: done https://github.com/DFE-Digital/teacher-training-api/pull/1579/commits/d1a210cdbf10e1cca1e2c95897167a09da4a88ac

**Comment by gpeng** [comment]
Discussed offline. This PR is update only. Show will be added in https://trello.com/c/X1EkaWrR/184-add-api-v2-contacts-id-show-endpoint. Create and destroy will be scoped for future work as will removing the now redundant individual contact elements on the provider response.

**Review by gpeng — APPROVED** [review]

File: app/controllers/api/v2/contacts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    class ContactsController < API::V2::ApplicationController
+      deserializable_resource :contact, class: API::V2::DeserializableContact
+
+      def update
+        authorize contact
+
+        if contact.update(contact_params)
+          render jsonapi: contact, status: :ok
+        else
+          render jsonapi_errors: contact.errors, status: :unprocessable_entity
```
Comment (defong) [inline app/controllers/api/v2/contacts_controller.rb:12]: Can we have the unhappy path tests please.

Comment (darokel) [reply app/controllers/api/v2/contacts_controller.rb:12]: done https://github.com/DFE-Digital/teacher-training-api/pull/1579/commits/996cb2eaadcae1d8aa7cb9fa0755d5682656c18b

**Review by defong — APPROVED** [review]

**Pull request merged**